File: templates/express-js/src/api/schemas/responses.ts
import { z } from "zod";
import { opportunityBaseSchema, oppSortingSchema } from "./models";
import { PaginationInfoSchema } from "./pagination";
import { oppDefaultFiltersSchema } from "./models";

// ############################################################################
// Base response schema
// ############################################################################

/** Response for a default operation */
export const defaultResponseSchema = z.object({
  /** The message */
  message: z.string(),
});

export type DefaultResponse = z.infer<typeof defaultResponseSchema>;

// ############################################################################
// Opportunity response schemas
// ############################################################################

/** A paginated list of opportunities */
export const opportunitiesListResponseSchema = defaultResponseSchema.extend({
  /** The list of opportunities */
  items: z.array(opportunityBaseSchema),
  /** The pagination details */
  paginationInfo: PaginationInfoSchema,
});

export type OpportunitiesListResponse = z.infer<typeof opportunitiesListResponseSchema>;

/** A paginated list of results from an opportunity search */
export const opportunitiesSearchResponseSchema = opportunitiesListResponseSchema.extend({
  /** The sorting details */
  sortInfo: oppSortingSchema,
  /** The filter details */
  filterInfo: oppDefaultFiltersSchema,
});

export type OpportunitiesSearchResponse = z.infer<typeof opportunitiesSearchResponseSchema>;

/** A single opportunity */
export const opportunityResponseSchema = defaultResponseSchema.extend({
  /** The opportunity */
  data: opportunityBaseSchema,
});

export type OpportunityResponse = z.infer<typeof opportunityResponseSchema>;

File: examples/ca-opportunity-example/src/common_grants/schemas/responses.py
"""Response models that use CAOpportunity for correct custom_fields serialization."""

from common_grants_sdk.schemas.pydantic import (
    OpportunitiesListResponse,
    OpportunitiesSearchResponse,
    OpportunityResponse,
)
from pydantic import Field

from common_grants.utils.opp_transform import CAOpportunity


class CAOpportunitiesListResponse(OpportunitiesListResponse):
    """List response with CA opportunities for correct custom_fields serialization."""

    items: list[CAOpportunity] = Field(
        ...,
        description="The list of opportunities",
    )


class CAOpportunitiesSearchResponse(OpportunitiesSearchResponse):
    """Search response with CA opportunities for correct custom_fields serialization."""

    items: list[CAOpportunity] = Field(
        ...,
        description="The list of opportunities",
    )


class CAOpportunityResponse(OpportunityResponse):
    """Single opportunity response with CA opportunity for correct custom_fields serialization."""

    data: CAOpportunity = Field(
        ...,
        description="The opportunity",
    )
